x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java

```diff
@@ -0,0 +1,595 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.stream.Collectors;
+
+public class TransportTermEnumAction extends HandledTransportAction<TermEnumRequest, TermEnumResponse> {
+
+    protected final ClusterService clusterService;
+    protected final TransportService transportService;
+    private SearchService searchService;
+    protected final IndexNameExpressionResolver indexNameExpressionResolver;
+
+    final String transportShardAction;
+    private final String shardExecutor;
+
+    @Inject
+    public TransportTermEnumAction(
+        // NodeClient client,
+        ClusterService clusterService,
+        TransportService transportService,
+        SearchService searchService,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(TermEnumAction.NAME, transportService, actionFilters, TermEnumRequest::new);
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        this.transportShardAction = actionName + "[s]";
+        // TODO new threadpool
+        this.shardExecutor = ThreadPool.Names.AUTO_COMPLETE;
+        this.searchService = searchService;
+        // this.client = client;
+
+        transportService.registerRequestHandler(
+            transportShardAction,
+            ThreadPool.Names.SAME,
+            ShardTermEnumRequest::new,
+            new ShardTransportHandler()
+        );
+
+    }
+
+    @Override
+    protected void doExecute(Task task, TermEnumRequest request, ActionListener<TermEnumResponse> listener) {
+        request.taskStartTimeMillis = task.getStartTime();
+        new AsyncBroadcastAction(task, request, listener).start();
+    }
+
+    protected ShardTermEnumRequest newShardRequest(int numShards, ShardRouting shard, TermEnumRequest request) {
+        final ClusterState clusterState = clusterService.state();
+        final Set<String> indicesAndAliases = indexNameExpressionResolver.resolveExpressions(clusterState, request.indices());
+        final AliasFilter aliasFilter = searchService.buildAliasFilter(clusterState, shard.getIndexName(), indicesAndAliases);
+        return new ShardTermEnumRequest(shard.shardId(), aliasFilter, request);
+    }
+
+    protected ShardTermEnumResponse readShardResponse(StreamInput in) throws IOException {
+        return new ShardTermEnumResponse(in);
+    }
+
+    protected GroupShardsIterator shards(ClusterState clusterState, TermEnumRequest request, String[] concreteIndices) {
+        final String routing = null;
+
+        // Remove any cold or frozen indices from the set of indices to be searched.
```

Comment:
We should rely on node's data-tier when selecting indices and not care about index settings. The goal is to   restrict the API to some data-tiers, we can add more flexibility in the future but I'd prefer that we start simple. Only the hot and warm tier should be queried.